src/dxtbx/model/goniometer.py

```diff
@@ -392,69 +392,69 @@ def imgCIF_H(cbf_handle):
-                depends_on.append(cbf_handle.get_value())
+                dependants[cbf_handle.get_value().decode()] = axis_name
             cbf_handle.next_row()
 
         # find the starting angles of each goniometer axis and figure out which one
         # is the scan axis (i.e. non-zero angle_increment)
         cbf_handle.find_category(b"diffrn_scan_axis")
         for i in range(cbf_handle.count_rows()):
             cbf_handle.find_column(b"axis_id")
-            axis_name = cbf_handle.get_value()
-            if axis_name.decode() not in axis_names:
+            axis_name = cbf_handle.get_value().decode()
+            if axis_name not in axis_vectors:
                 cbf_handle.next_row()
                 continue
             cbf_handle.find_column(b"angle_start")
             axis_angle = float(cbf_handle.get_value())
             cbf_handle.find_column(b"angle_increment")
             increment = float(cbf_handle.get_value())
-            angles.append(axis_angle)
-            if abs(increment) > 0:
-                assert (
-                    scan_axis is None
-                ), "More than one scan axis is defined: not currently supported"
-                scan_axis = flex.first_index(axis_names, axis_name)
+            angles[axis_name] = axis_angle
+            if increment != 0:
+                if scan_axis:
+                    raise ValueError(
+                        "More than one scan axis is defined: not currently supported."
+                    )
+                scan_axis = axis_name
             cbf_handle.next_row()
-        assert axes.size() == angles.size()
-        if scan_axis is None:
-            # probably a still shot -> scan axis arbitrary as no scan
-            scan_axis = 0
+        if not len(axis_vectors) == len(angles):
+            raise ValueError(
+                "The number of goniometer axes specified in the 'axis' category of the "
+                "raw data does not match the number specified in the "
+                "'diffrn_scan_axis' category."
+            )
 
         # figure out the order of the axes from the depends_on values
-        order = flex.size_t()
-        for i in range(axes.size()):
-            if depends_on[i] == ".":
-                o = 0
-            else:
-                o = flex.first_index(axis_names, depends_on[i]) + 1
-            assert o not in order
-            order.append(o)
+        ordered_axes = []
+        axis = "."
+        while axis := dependants.get(axis):
+            ordered_axes.append(axis)
 
         # multi-axis gonio requires axes in order as viewed from crystal to gonio base
         # i.e. the reverse of the order we have from cbf header
-        order = order.reversed()
-        axes = axes.select(order)
-        angles = angles.select(order)
-        axis_names = axis_names.select(order)
-        scan_axis = axes.size() - scan_axis - 1
+        ordered_axes = ordered_axes[::-1]
```

Comment:
This is preferable to `reversed()`?

Reply:
Reversed returns an iterator, so you can't use it twice.

Reply:
👍 so yes it is